File: problem_solving/hash_table/py/1244_design_leaderboard.py
'''
https://leetcode.com/problems/design-a-leaderboard/

t: N* O(log K)
s: O(K)
'''
from collections import defaultdict
import heapq

class Leaderboard:

    def __init__(self):
        self.scores = defaultdict(int)

    def addScore(self, playerId: int, score: int) -> None:
        self.scores[playerId] += score

    def top(self, K: int) -> int:
        #return sum(sorted(self.scores.values(), reverse=True)[:K])
        return sum(heapq.nlargest(K, self.scores.values()))


    def reset(self, playerId: int) -> None:
        self.scores[playerId] = 0


# Your Leaderboard object will be instantiated and called as such:
leaderboard = Leaderboard()
leaderboard.addScore(1, 73)  # leaderboard = {1: 73}
leaderboard.addScore(2, 56)  # leaderboard = {1: 73, 2: 56}
print(leaderboard.top(1))          # returns 73
leaderboard.reset(1)        # leaderboard = {1: 0, 2: 56}
leaderboard.addScore(2, 51) # leaderboard = {1: 0, 2: 107}
print(leaderboard.top(2))          # returns 107 (107 + 0)

File: problem_solving/hash_table/cpp/1244_design_leaderboard.cpp
/*
https://leetcode.com/problems/design-a-leaderboard/

std::unordered<int, int> playerid_score_map
in top method, create a min heap..  if size() > K, pop()

unordered_map => 
*/

#include<queue>
#include<iostream>
#include "../../../cpp/common_header.h"
#include "../../../cpp/utils/utils.h"
#include<stack>
#include<unordered_map>


class Leaderboard {
    private:
    std::unordered_map<int, int> scores_map;
public:
    Leaderboard() {
        
    }
    
    void addScore(int playerId, int score) {
        scores_map[playerId] += score;
    }
    
    int top(int K) {
        int sum_k = 0;
        std::priority_queue<int, std::vector<int>, std::greater<int>> heap;
        for(const auto& [key, value] : scores_map) {
            heap.push(value);
            if(heap.size() > K) {
                heap.pop();
            }
        }
        while(!heap.empty()) {
            sum_k += heap.top();
            heap.pop();
        }
        return sum_k;
    }
    
    int top_players(int K) {
        int sum_k = 0;
        std::priority_queue< std::pair<int,int>, std::vector< std::pair<int,int> >, std::greater<std::pair<int,int>>> heap;
        for(const auto& pair : scores_map) {
            heap.push({pair.second, pair.first});
            if(heap.size() > K) {
                heap.pop();
            }
        }
        while(!heap.empty()) {
            const auto& pr = heap.top();
            sum_k += pr.first;
            int Id = pr.second;
            heap.pop();
        }
        return sum_k;
    }

    void reset(int playerId) {
        scores_map[playerId] = 0;
    }
};

int main() {
/**
 * Your Leaderboard object will be instantiated and called as such:
 * Leaderboard* obj = new Leaderboard();
 * obj->addScore(playerId,score);
 * int param_2 = obj->top(K);
 * obj->reset(playerId);
 */
    Leaderboard* obj = new Leaderboard();
    int playerId = 1;
    int score = 10;
    int K = 4;
    obj->addScore(playerId,score);
    int param_2 = obj->top(K);
    obj->reset(playerId);

}